cdbot/cogs/admin.py

```diff
@@ -89,6 +92,32 @@ async def on_member_join(self, member: Member):
             # assign placeholder nickname
             await member.edit(nick=PLACEHOLDER_NICKNAME)
 
+    @Cog.listener()
+    async def on_message(self, message):
+        def check(m):
+            return m.author == message.author and (datetime.datetime.utcnow() - m.created_at).seconds < 4
+
+        def who(m):
+            return m.author == message.author
+
+        if not message.author.bot:
+            # Checks if message contains banned word.
+            if any([word in message.content for word in BANNED_WORDS]):
+                await message.delete()
+                await message.channel.send(f"{message.author.mention} Watch your language!", delete_after=5)
+            # Checks if message contains banned domain.
+            if any([word in message.content for word in BANNED_DOMAINS]):
+                await message.delete()
+                await message.channel.send(f"{message.author.mention} No invite links!", delete_after=5)
+            # Checks if message contains mass mention.
+            if len(message.mentions) > 8:
+                await message.delete()
+                await message.channel.send(f"{message.author.mention} Don't spam mentions!", delete_after=5)
+            # Checks if message was sent too quickly.
+            if len(list(filter(lambda m: check(m), self.client.cached_messages))) >= 4:
+                await message.channel.purge(limit=4, check=who)
```

Comment:
The magic number `4` has just been used twice here. For starters, it should be defined in a config somewhere. Secondly, because you're checking if it's `>= 4`, rather than `== 4`, we need to handle the case where it's `> 4` as well. Additionally, messages will come into the bot faster than the deletion event for the purge. This means that if I spam 5 messages, the first 4 will be caught by the check, and deleted, but then messages 2 though 5 will also match the check, as the deletion response for 1 though 4 has not yet arrived. This causes the bot to double-warn and attempt to delete messages 2 though 4 twice.

Furthermore, `limit` is not the number of deleted messages. As the docs say:

> limit (Optional[int]) – The number of messages to search through. This is not the number of messages that will be deleted, though it can be.

This means this filter can be trivially bypassed by having two or more members spam in unison. The bot will attempt to delete the most recent 4 messages, however if the check fails for all 4, it will delete none and will not continue searching.